dashboard/test/models/pd/application/teacher1920_application_test.rb

```diff
@@ -1004,6 +1004,27 @@ class Teacher1920ApplicationTest < ActiveSupport::TestCase
       assert_equal 'Complete - Yes', application.reload.principal_approval_state
     end
 
+    test 'require assigned workshop for some statuses when emails sent by system' do
+      statuses = ['accepted_no_cost_registration', 'registration_sent']
+      partner = build :regional_partner, applications_decision_emails: RegionalPartner::SENT_BY_SYSTEM
+      workshop = create :pd_workshop, location_address: 'Address', sessions_from: Date.today, num_sessions: 1
+      application = create :pd_teacher1920_application, {
+        regional_partner: partner
+      }
+
+      statuses.each do |status|
+        application.status = status
+        refute application.valid?
+        assert_equal ["#{status} requires workshop to be assigned"], application.errors.messages[:status]
+      end
+
+      application.pd_workshop_id = workshop.id
+      statuses.each do |status|
+        application.status = status
+        assert application.valid?
+      end
```

Comment:
Good - just add one thing testing that you can assign other statuses to the application without having to assign a workshop_id